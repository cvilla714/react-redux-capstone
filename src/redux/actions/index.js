import axios from 'axios';
import {
  CLEAR_USERS,
  GET_REPOS,
  GET_USER,
  SEARCH_USERS,
  SET_LOADING,
  LOGS_ERROR,
} from './types';

export const setLoading = () => ({
  type: SET_LOADING,
});

export const searchUsers = (text) => async (dispatch) => {
  try {
    setLoading();

    const res = await axios.get(
      `https://api.github.com/search/users?q=${text}`,
      {
        headers: {
          Authorization: `token ${process.env.REACT_APP_GITHUB_TOKEN}`,
        },
      },
    );
    dispatch({
      type: SEARCH_USERS,
      payload: res.data.items,
    });
  } catch (error) {
    dispatch({
      type: LOGS_ERROR,
      payload: error.respose.statusText,
    });
  }
};

export const getUser = (username) => async (dispatch) => {
  try {
    setLoading();

    const res = await axios.get(`https://api.github.com/users/${username}`, {
      headers: {
        Authorization: `token ${process.env.REACT_APP_GITHUB_TOKEN}`,
      },
    });

    dispatch({
      type: GET_USER,

      payload: res.data,
    });
  } catch (error) {
    dispatch({
      type: LOGS_ERROR,
      payload: error.respose.statusText,
    });
  }
};

export const getUserRepos = (username) => async (dispatch) => {
  try {
    setLoading();

    const res = await axios.get(
      `https://api.github.com/users/${username}/repos?per_page=5&sort=created:asc`,
      {
        headers: {
          Authorization: `token ${process.env.REACT_APP_GITHUB_TOKEN}`,
        },
      },
    );

    dispatch({
      type: GET_REPOS,
      payload: res.data,
    });
  } catch (error) {
    dispatch({
      type: LOGS_ERROR,
      payload: error.respose.statusText,
    });
  }
};

export const clearUsers = () => ({
  type: CLEAR_USERS,
});
