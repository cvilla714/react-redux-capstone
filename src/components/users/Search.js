import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { searchUsers, clearUsers } from '../../actions/index';
import setAlert from '../../actions/alertActions';

const Search = ({
  searchUsers, clearUsers, setAlert, users,
}) => {
  const [text, setText] = useState('');

  const onSubmit = (e) => {
    e.preventDefault();
    if (text === '') {
      setAlert('Please enter something', 'light');
    } else {
      searchUsers(text);
      setText('');
    }
  };

  const onChange = (e) => {
    setText(e.target.value);
  };

  return (
    <div>
      <form onSubmit={onSubmit} className="form">
        <input
          type="text"
          name="text"
          placeholder="Search User..."
          value={text}
          onChange={onChange}
        />
        <input
          type="submit"
          value="Search"
          className="btn btn-dark btn-block"
        />
      </form>
      {users.length > 0 && (
        <button
          className="btn btn-light btn-block"
          type="button"
          onClick={clearUsers}
        >
          Clear
        </button>
      )}
    </div>
  );
};

Search.propTypes = {
  searchUsers: PropTypes.func.isRequired,
  clearUsers: PropTypes.func.isRequired,
  setAlert: PropTypes.func.isRequired,
  users: PropTypes.arrayOf(Array).isRequired,
};

const mapStateToProps = (state) => ({
  users: state.github.users,
  alert: state,
});

export default connect(mapStateToProps, { searchUsers, clearUsers, setAlert })(
  Search,
);
