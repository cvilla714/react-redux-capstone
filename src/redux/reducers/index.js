import { combineReducers } from 'redux';
import githubReducer from './githubReducer';
import alertReducer from './alertReducer';
import filterReducer from './filterReducer';

export default combineReducers({
  github: githubReducer,
  alert: alertReducer,
  filter: filterReducer,
});
