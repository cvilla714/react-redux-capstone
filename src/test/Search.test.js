import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { create } from 'react-test-renderer';
import { createStore, applyMiddleware } from 'redux';
import combineReducers from '../reducers';
import Search from '../containers/users/Search';

test('Search renders correctly', () => {
  const store = createStore(combineReducers, applyMiddleware(thunk));
  const tree = create(
    <Provider store={store}>
      <Search />
    </Provider>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
