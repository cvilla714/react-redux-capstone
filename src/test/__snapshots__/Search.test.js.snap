// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders correctly 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      name="text"
      onChange={[Function]}
      placeholder="Search User..."
      type="text"
      value=""
    />
    <input
      className="btn btn-dark btn-block"
      type="submit"
      value="Search"
    />
  </form>
</div>
`;
